### PR TITLE
[LIBCLOUD-938] New compute driver for UpCloud

### PR DESCRIPTION
**Pull request by mlackman**
### Description

UpCloud compute driver. Will be done against Upcloud's 1.2.4 api version
UpCloud's api documentation can be found from:
https://www.upcloud.com/api/1.2.4/

### Status
done, ready for review

### Checklist (tick everything that applies)

- [x] [Code linting](http://libcloud.readthedocs.org/en/latest/development.html#code-style-guide) (required, can be done after the PR checks)
- [x] Documentation
- [x] [Tests](http://libcloud.readthedocs.org/en/latest/testing.html)
- [x] [ICLA](http://libcloud.readthedocs.org/en/latest/development.html#contributing-bigger-changes) (required for bigger changes)

### CODE REVIEW

**Comment by tonybaloney** [comment]
cool, ok thanks for the contribution! there's a lot to review here so it might take a couple of days for this to get reviewed.

**Comment by iler** [comment]
Hi @tonybaloney, any progress with this? :)

**Comment by mlackman** [comment]
Hi @tonybaloney, yeah sorry for such a big commit. Is there anything I can do to help the review?

File: libcloud/common/upcloud.py
Comment (tonybaloney) [inline libcloud/common/upcloud.py]: This should inherit from one of the Libcloud Exception types

File: libcloud/common/upcloud.py
Comment (tonybaloney) [inline libcloud/common/upcloud.py]: New line after the quote marks please http://libcloud.readthedocs.io/en/latest/development.html#docstring-conventions

File: libcloud/common/upcloud.py
Comment (tonybaloney) [inline libcloud/common/upcloud.py:94]: Please add docstrings to explain the arguments http://libcloud.readthedocs.io/en/latest/development.html#docstring-conventions

**Review by tonybaloney — CHANGES_REQUESTED** [review]

**Comment by tonybaloney** [comment]
hi just a couple of things to edit. please also add a doc page

File: libcloud/common/upcloud.py
Comment (mlackman) [reply libcloud/common/upcloud.py]: Ok, the UpcloudTimeoutException is used when server state is not getting correct state in timely fashion. Its not coming from the REST interface, the timeout is figured out in the implementation. I checked the libcloud's exception types and I did not find suitable exception type except the generic LibCloudError. Other exception types were for errors coming from the REST api. 

What is your take on using LibCloudError?

File: docs/compute/drivers/upcloud.rst
Comment (pquentin) [inline docs/compute/drivers/upcloud.rst]: Can we be more objective? You can look at the pages from other providers. Thanks!

Comment (iler) [reply docs/compute/drivers/upcloud.rst]: Revised the wording there and updated the PR accordingly :)

File: docs/compute/drivers/upcloud.rst
Comment (pquentin) [inline docs/compute/drivers/upcloud.rst]: Thank you for the revised wording!

s/data centres/data centers/

File: libcloud/compute/drivers/upcloud.py
Comment (pquentin) [inline libcloud/compute/drivers/upcloud.py]: Can you please define rename this to `NODE_STATE_MAP` and define it in `UpcloudDriver` for better consistency with other drivers?

File: libcloud/common/upcloud.py
Comment (pquentin) [inline libcloud/common/upcloud.py]: Since this is a non-standard method name, please add the `ex_` prefix. And since it does not modify state, I think `ex_get_node_state` is better, but that's up to you.

**Review by pquentin — CHANGES_REQUESTED** [review]
There are only two small changes left before merging this. Thank you for bearing with us!

File: libcloud/common/upcloud.py
Comment (mlackman) [reply libcloud/common/upcloud.py]: Yeah the get_node_state is a better name for it, but is the ex_ needed here because this is a helper class and not part of the UpcloudDriver api.

Comment (pquentin) [reply libcloud/common/upcloud.py]: Oh, right. Not needed, no, sorry.

**Review by pquentin — APPROVED** [review]

**Comment by pquentin** [comment]
Will merge tomorrow morning, thanks.

**Comment by mlackman** [comment]
@pquentin Do I follow this instruction from here https://libcloud.readthedocs.io/en/latest/development.html. So I need to squash the commits and make a patch?

**Comment by pquentin** [comment]
@mlackman The PR has been merged in trunk, see https://github.com/apache/libcloud/commits. Since the GitHub repository is only a mirror, I pushed in into the Apache repository, which is why the PR is shown as closed.

I'm a new libcloud committer. I guess I should have squashed the commits, but I... forgot about it. Too late now, awnyways.

**Pull request closed without merging**
